fix: filesystemStore fails to prepend target file hashes on Windows

torin-carey opened the pull request:
Fixes #273
Release Notes: Fix commit for Windows when using filesystemStore.

**Types of changes**:
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

**Description of the changes being introduced by the pull request**:

The `filesystemStore` `LocalStore` implementation conflates paths and filepaths,
which cause issues on systems where they take different formats (Windows).
This is most noticeable when running a repository with consistent snapshots, where committed targets will not have a prepended file hash, since the `filesystemStore` implementation mistakes them for metadata files, rather than targets:
```go
// local_store.go
func (f *fileSystemStore) Commit(consistentSnapshot bool, versions map[string]int, hashes map[string]data.Hashes) error {
	isTarget := func(path string) bool {
		return strings.HasPrefix(path, "targets/")
	}
	copyToRepo := func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		if info.IsDir() || !info.Mode().IsRegular() {
			return nil
		}
		rel, err := filepath.Rel(f.stagedDir(), path)
		if err != nil {
			return err
		}

		var paths []string
		if isTarget(rel) {
			paths = computeTargetPaths(consistentSnapshot, rel, hashes)
		} else {
			paths = computeMetadataPaths(consistentSnapshot, rel, versions)
		}
		// ...
	}
	// ...
	if err := filepath.Walk(f.stagedDir(), copyToRepo); err != nil {
		return err
	}
	// ...
```
Because of this, clients will fail to find the target file from the repository.

**Please verify and check that the pull request fulfills the following requirements**:

- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

znewman01 commented:
This looks great, thanks!

2 things:

1. can you add Windows to the test matrix in `.github/workflows/tests.yml`? Should just be a matter of adding `windows-latest` here: https://github.com/theupdateframework/go-tuf/blob/3dfbeb242bfcf3964f4ccf3327d36e165b3c2384/.github/workflows/tests.yml#L20 

2. Test failure. I'm a little perplexed as to why this happens specifically on 1.17...it may be unrelated but I haven't seen it fail this way before.

rdimitrov commented:
@torin-carey - It's important to have consistent support for Windows too, so thank you for helping with that! 💯 

Perhaps it needs a bit more polishing though, because CI failed not only for 1.17, but for the whole set of test cases (it cancels the rest when one fails). You can try debugging these issues by running the test suite locally via `go test ./...` 

Also, as @znewman01 suggested, adding the Windows environment to the CI config would be useful to make sure we support it continuously 👍 

Partially related - #131, #79, #80

torin-carey commented:
I've sorted the failing tests (bug in original commit causing files to not be deleted after removing target) and all of the go tests are passing on Linux and Windows. I've also replaced the use of `filepath` with `path` in `(*DelegatedRole).MatchesPath` since the escape semantics of `filepath.Match` and `path.Match` are different between Windows and *nix.

I've also added `windows-latest` to the workflows test matrix but haven't been able to test this with workflows in a separate repo, so :crossed_fingers:

znewman01 reviewed (commented):
Thank you very much! This looks fine to me barring the test failures (which might just be lint related?)

znewman01 commented:
Looking at the last test failures: I see https://github.com/theupdateframework/go-tuf/blob/ad96eca0239ec2cc9b6e408fbe42b2f9e9d6b1dd/client/interop_test.go#L47

It looks like one of those is reading `\r\n`. Maybe related to https://github.com/actions/checkout/issues/135 . I think that we could either turn `autoclrf` off in a test environment, or make the `computeHashes` test function normalize line endings. Maybe the latter is better so that tests will actually work in a windows source checkout.

torin-carey commented:
Thanks for reviewing!
> It looks like one of those is reading `\r\n`. Maybe related to [actions/checkout#135](https://github.com/actions/checkout/issues/135) . I think that we could either turn `autocrlf` off in a test environment, or make the `computeHashes` test function normalize line endings. Maybe the latter is better so that tests will actually work in a windows source checkout.

Thanks for finding this, the line endings do seem to be the only issue affecting the windows tests. I don't think it would suffice to just change the reader, since the mangling of the metadata files by git really does infringe on their integrity and I don't think we should try and accommodate that. Adding `.gitattributes` to either mark all of the testdata as binary or have `eol=lf` I think would be the right approach.

The linter errors also seem to be about line endings, since `gofmt` wants to correct `\r\n => \n` (see https://github.com/golangci/golangci-lint/issues/580). Maybe we should add a project-wide `*.go eol=lf` git attribute?

Another problem I've noticed running git on windows locally is that git for windows by default does not create symlinks, causing some of the test cases to fail where test case metadata files are de-duplicated through links. Github workflows seems to enable `core.symlinks` so this shouldn't be a problem for the CI, but may be for users. Should we remove the symlinks or just leave it as-is?

torin-carey commented on client/testdata/go-tuf/generator/generator.go:
```diff
@@ -29,8 +30,37 @@ func assertNoError(err error) {
 }
 
 func copyRepo(src string, dst string) {
-	cmd := exec.Command("cp", "-r", src, dst)
```
Thanks for the feedback, I've added a docstring to it.

znewman01 commented:
>  I don't think it would suffice to just change the reader, since the mangling of the metadata files by git really does infringe on their integrity and I don't think we should try and accommodate that. Adding .gitattributes to either mark all of the testdata as binary or have eol=lf I think would be the right approach.

You've convinced me, I think that's correct.

> The linter errors also seem to be about line endings, since gofmt wants to correct \r\n => \n (see https://github.com/golangci/golangci-lint/issues/580). Maybe we should add a project-wide *.go eol=lf git attribute?

SGTM

>Another problem I've noticed running git on windows locally is that git for windows by default does not create symlinks, causing some of the test cases to fail where test case metadata files are de-duplicated through links. Github workflows seems to enable core.symlinks so this shouldn't be a problem for the CI, but may be for users. Should we remove the symlinks or just leave it as-is?

Good find. I think we can address this later; I'll file an issue. The symlinks are all consistent snapshot-related so I think they're somewhat deliberate.

TBH I'd prefer to keep the symlinks and fix this with documentation. If we *must* remove them, we'd need some automation around making sure the files stay in sync.

joshuagl commented:
> Thanks for reviewing!
> 
> > It looks like one of those is reading `\r\n`. Maybe related to [actions/checkout#135](https://github.com/actions/checkout/issues/135) . I think that we could either turn `autocrlf` off in a test environment, or make the `computeHashes` test function normalize line endings. Maybe the latter is better so that tests will actually work in a windows source checkout.
> 
> Thanks for finding this, the line endings do seem to be the only issue affecting the windows tests. I don't think it would suffice to just change the reader, since the mangling of the metadata files by git really does infringe on their integrity and I don't think we should try and accommodate that. Adding `.gitattributes` to either mark all of the testdata as binary or have `eol=lf` I think would be the right approach.

💯 this is what we did for python-tuf

znewman01 commented:
1 more annoying thing, sorry: https://github.com/theupdateframework/go-tuf/pull/319

(You'll need to rebase and do something like: https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md before you push again)

znewman01 commented:
@torin-carey anything the maintainers can do on our end to help out with this?

(No urgency, just want to make sure you're not blocked)

torin-carey commented:
Hi @znewman01, I apologise for the delayed response.

The current state of the PR works for me with both `golangci-lint` and go test/check through a similar git config to workflows. Hopefully workflows agrees :crossed_fingers:.  Commits have also been signed off for DCO.

joshuagl reviewed (commented):
Thank you so much for your work on this! I'm looking forward to having Windows working and tested :-)

The PR looks great. I just have one small question about the gitattributes file.

torin-carey commented on .gitattributes:
```diff
@@ -0,0 +1,5 @@
+*.go eol=lf
+
+# testdata should not be mangled by git
+*.json binary
```
Thanks for having a look at the code! I felt that since mangling of any kind is bad for file integrity, I tended towards marking them as binary.
We do expect all test files in the repo to be using LF line endings so mangling shouldn't be an issue anyway.
Would `eol=lf` be preferable to `binary`?

znewman01 replied:
+1, @torin-carey: I think that we should treat the test data as binary for exactly that reason

joshuagl replied:
Yes, agreed. I think we should change python-tuf to treat the test data as binary too. Thanks both.

znewman01 commented:
Oh, I'm taking @rdimitrov off as reviewer because he's currently on leave.

@joshuagl can you review for approval?

joshuagl reviewed (approved):
LGTM, thanks so much @torin-carey !

torin-carey commented:
Thanks for reviewing and approving!
